modules/contact/test/tests/tan-pen-and-scaling/bouncing-block-tan-pen.i

```diff
@@ -0,0 +1,113 @@
+starting_point = 2e-1
+offset = 1e-2
+diffusivity = 1e3
+
+[GlobalParams]
+  displacements = 'disp_x disp_y'
+  diffusivity = ${diffusivity}
+[]
+
+[Mesh]
+  file = long-bottom-block-no-lower-d-coarse.e
+[]
+
+[Variables]
+  [./disp_x]
+  [../]
+  [./disp_y]
+  [../]
+[]
+
+[ICs]
+  [./disp_y]
+    block = 2
+    variable = disp_y
+    value = ${fparse starting_point + offset}
+    type = ConstantIC
+  [../]
+[]
+
+[Kernels]
+  [./disp_x]
+    type = MatDiffusion
+    variable = disp_x
+  [../]
+  [./disp_y]
+    type = MatDiffusion
+    variable = disp_y
+  [../]
+[]
+
+[Contact]
+  [leftright]
+    slave = 10
+    master = 20
+    model = coulomb
+    formulation = tangential_penalty
+    penalty = ${diffusivity}
+    system = constraint
+    friction_coefficient = 0.4
+  []
+[]
+
+[BCs]
+  [./botx]
+    type = DirichletBC
+    variable = disp_x
+    boundary = 40
+    value = 0.0
+  [../]
+  [./boty]
+    type = DirichletBC
+    variable = disp_y
+    boundary = 40
+    value = 0.0
+  [../]
+  [./topy]
+    type = FunctionDirichletBC
+    variable = disp_y
+    boundary = 30
+    function = '${starting_point} * cos(2 * pi / 40 * t) + ${offset}'
+  [../]
+  [./leftx]
+    type = FunctionDirichletBC
+    variable = disp_x
+    boundary = 50
+    function = '1e-2 * t'
+  [../]
+[]
+
+[Executioner]
+  type = Transient
+  end_time = 100
+  dt = 5
+  dtmin = 2.5
+  solve_type = 'PJFNK'
+  petsc_options = '-snes_converged_reason -ksp_converged_reason'
+  petsc_options_iname = '-pc_type -pc_hypre_type -mat_mffd_err -ksp_gmres_restart'
+  petsc_options_value = 'hypre    boomeramg      1e-5          200'
+  l_max_its = 200
+  nl_max_its = 20
+  line_search = 'none'
+  automatic_scaling = true
+  verbose = true
+  scaling_group_variables = 'disp_x disp_y'
```

Comment:
How important is it to group the variables for scaling?  Is this something we would want to do even for non-contact problems?

Reply:
I think so. I have this input file from Al:

<details>

<summary>Al input</summary>

```
[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
  volumetric_locking_correction = true
  # scaling = 1e0
[]

[Mesh]
  type = GeneratedMesh
  dim = 3
  nx = 5
  ny = 1
  nz = 1
  xmin = 0.0
  xmax = 4.0
  ymin = 0.0
  ymax = 0.5
  zmin = 0.0
  zmax = 1.0
  elem_type = HEX8
  uniform_refine = 1
[]

[Variables]
  [disp_x]
    # scaling = 3.25447e-06
  []
  [disp_y]
    # scaling = 2.29038e-06
  []
  [disp_z]
    # scaling = 3.60462e-06
  []
[]

[Modules/TensorMechanics/Master]
  [./all]
    strain = FINITE
    generate_output = 'stress_xx stress_xy stress_yy strain_xx strain_xy strain_yy'
  [../]
[]

[Functions]
  [./loading_func]
    type = PiecewiseLinear
    x = '0.  5.'
    y = '0. 50.0'
  [../]
[]

[BCs]
  [./free_end_moment]
    type = Pressure
    variable = disp_y
    component = 0
    boundary = right
    factor = 1
    function = loading_func
  [../]
  [./FixedCenterLineX]
    type = DirichletBC
    variable = disp_x
    boundary = left
    value = 0.0
  [../]
  [./FixedCenterLineY]
    type = DirichletBC
    variable = disp_y
    boundary = left
    value = 0.0
  [../]
  [./FixedCenterLineZ]
    type = DirichletBC
    variable = disp_z
    boundary = left
    value = 0.0
  [../]
[]

[Materials]
  [./elasticity_tensor]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 1e6
    poissons_ratio = 0.3
  [../]
  [./stress]
    type = ComputeFiniteStrainElasticStress
  [../]
[]

[Preconditioning]
  [./pc]
    type = SMP
    full = true
  [../]
[]

[Executioner]
  type = Transient
  solve_type = 'NEWTON'
  petsc_options_iname = '-ksp_gmres_restart'
  petsc_options_value = '200'

  l_max_its = 200

  nl_max_its = 20

  start_time = 0.0
  dt = 1
  dtmin = 1
  end_time = 5
  num_steps = 1
[]

[Postprocessors]
  [./num_lin_it]
    type = NumLinearIterations
  [../]
  [./num_nonlin_it]
    type = NumNonlinearIterations
  [../]
  [./tot_lin_it]
    type = CumulativeValuePostprocessor
    postprocessor = num_lin_it
  [../]
  [./tot_nonlin_it]
    type = CumulativeValuePostprocessor
    postprocessor = num_nonlin_it
  [../]
  [./alive_time]
    type = PerfGraphData
    section_name = Root
    data_type = TOTAL
  [../]
  [./max_beam_deflection]
    type = NodalMaxValue
    variable = disp_y
    boundary = 'right'
  [../]
[]

[Outputs]
  exodus = true
  csv = true
  print_linear_residuals = true
[]
```

</details>

where if I do normal automatic scaling the residual is 2.718520e-12 after 11 nonlinear iterations. (initial residual of 2.147227e-06). The automatic scaling factors are:
```
Automatic scaling factors:
  disp_x: 1.62724e-06
  disp_y: 1.14519e-06
  disp_z: 1.80231e-06
```
If I group them together, the residual after 11 nonlinear iterations is 8.083608e-15 with a single scaling factor of `1.14519e-06`.

I think that for the same physics it might be important to apply the same scaling factors to preserve the physics information in the structure of the matrix. In my initial concept of automatic scaling I thought we would scale every matrix row individually such that each row would have a maximum absolute value of unity, but @jedbrown pointed out that this would render preconditioners like multigrid unusable. So now we scale by variable. Perhaps extending it to group similar variables also helps the preconditioner perform optimally.

Reply:
One can recover, but you'd need `MatSetNearNullSpace` so AMG knows what it needs to approximate on the coarse grid. If you use preconditioned residual norm, these scaling problems roughly go away (modulo numerical stability).